# Policy Prediction



<b>Library Imports</b> :

In [1]:
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

'''
axis : {0 or ‘index’, 1 or ‘columns’}, default 0
0 or ‘index’ for row-wise, 1 or ‘columns’ for column-wise

'''

'\naxis : {0 or ‘index’, 1 or ‘columns’}, default 0\n0 or ‘index’ for row-wise, 1 or ‘columns’ for column-wise\n\n'

<b>Read Data</b> :

In [2]:
train = pd.read_csv("Data/train.csv")
train = pd.read_csv("Data/train_short.csv")
test = pd.read_csv("Data/test.csv")
test_session = pd.read_csv("Data/test_session_history.csv")


<b>Explore Training Data</b>:

In [3]:
train.head()

customer_ID  shopping_pt  record_type  day   time state  location  \
0       140813            9            1    2  13:02    PA     37390   
1       140815            6            1    3  13:41    FL     33166   
2       140809            7            1    4  14:45    PA     39166   
3       140811            6            1    1  12:42    RI     38272   
4       281617            7            1    2   9:47    IA     43903   

   group_size  homeowner  car_age car_value  risk_factor  age_oldest  \
0           1          0       13         e          NaN          24   
1           2          0        2         e          2.0          52   
2           1          1       20         d          NaN          51   
3           1          0       17         c          3.0          24   
4           1          0       12         e          NaN          55   

   age_youngest  married_couple  C_previous  duration_previous  policy  cost  
0            24               0         1.0                5.0       1   675  
1            21               0         1.0                6.0       4   653  
2            51               0         3.0                1.0       1   620  
3            24               0         3.0                8.0       3   661  
4            55               0         4.0                4.0       3   595

<b>Data Preparation and Cleaning</b> :

`Step 1` : Remove the default index added to the Data Frame

We can set the customer_ID as the index as this uniquely identifies the customer in the dataframe/dataset.
To do this we can use the Pandas <b><i>set_index</i></b> function.

<u>To find missing values</u> : use the Pandas count function.

The count function returns Series with number of non-NA/null observations over requested axis. Works with non-floating point data as well (detects NaN and None).

In the step below we can see that there are 67663 rows in all columns except - risk_factor, C_previous and duration_previous. This is because these columns have missing/NaN values.

In [4]:
train.count(axis=0, level=None, numeric_only=False)

customer_ID          67663
shopping_pt          67663
record_type          67663
day                  67663
time                 67663
state                67663
location             67663
group_size           67663
homeowner            67663
car_age              67663
car_value            67663
risk_factor          43618
age_oldest           67663
age_youngest         67663
married_couple       67663
C_previous           67103
duration_previous    67103
policy               67663
cost                 67663
dtype: int64

An alternate to the count function is the info function, which gives the concise summary of a DataFrame as below.

In [5]:
train.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67663 entries, 0 to 67662
Data columns (total 19 columns):
customer_ID          67663 non-null int64
shopping_pt          67663 non-null int64
record_type          67663 non-null int64
day                  67663 non-null int64
time                 67663 non-null object
state                67663 non-null object
location             67663 non-null int64
group_size           67663 non-null int64
homeowner            67663 non-null int64
car_age              67663 non-null int64
car_value            67663 non-null object
risk_factor          43618 non-null float64
age_oldest           67663 non-null int64
age_youngest         67663 non-null int64
married_couple       67663 non-null int64
C_previous           67103 non-null float64
duration_previous    67103 non-null float64
policy               67663 non-null int64
cost                 67663 non-null int64
dtypes: float64(3), int64(13), object(3)
memory usage: 9.8+ MB


To drop the NaN values, we can use the Pandas 'dropna' function

The dropna function returns objects with labels on given axis omitted where alternately any or all of the data are missing

In [6]:
columns = train.columns

In [7]:
train_new = train.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)

In [8]:
train_new.count(axis=0, level=None, numeric_only=False)

customer_ID          43349
shopping_pt          43349
record_type          43349
day                  43349
time                 43349
state                43349
location             43349
group_size           43349
homeowner            43349
car_age              43349
car_value            43349
risk_factor          43349
age_oldest           43349
age_youngest         43349
married_couple       43349
C_previous           43349
duration_previous    43349
policy               43349
cost                 43349
dtype: int64

This shows that all the NaN's have been deleted. The number of rows decreased from 67663 to 43349, i.e, about 35%..!!




<u>Drop Duplicates</u> : Since every row has a set of features that represent the customer

In [9]:
del train['customer_ID']
train.set_index('shopping_pt', drop=True, inplace=True)